rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,253 @@
+    ```
+
+2.  Deploy Vector!
+
+    Now that you have your custom `ConfigMap` ready it's time to deploy
+    Vector. To ensure Vector is isolated and has the necessary permissions
+    we must create a `namespace`, `ServiceAccount`, `ClusterRole`, and
+    `ClusterRoleBinding`:
+
+    ```bash
+    kubectl create namespace logging
+    kubectl create -f vector-service-account.yaml
+    kubectl create -f vector-role.yaml
+    kubectl create -f vector-role-binding.yaml
+    kubectl create -f vector-configmap.yaml
+    kubectl create -f vector-daemonset.yaml
+    ```
+
+    * *See [outstanding questions 3, 4, 5, 6, and 7](#outstanding-questions).*
+
+    That's it!
+
+#### Helm Interface
+
+TODO: fill in
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+2. [Fluentd k8s integration]
+3. [LogDNA k8s integration]
+4. [Honeycomb integration]
+3. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+4. [Influx Helm charts]
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community driven
+   integrations.
+
+## Outstanding Questions
+
+1. What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].
+1. What is the best to avoid Vector from ingesting it's own logs? I'm assuming
+   that my [`kubectl` tutoria](#kubectl-interface) handles this with namespaces?
+   We'd just need to configure Vector to excluse this namespace?
+1. I've seen two different installation strategies. For example, Fluentd offers
+   a [single daemonset configuration file][fluentd_daemonset] while Fluentbit
+   offers [four separate configuration files][fluentbit_installation]
+   (`service-account.yaml`, `role.yaml`, `role-binding.yaml`, `configmap.yaml`).
+   Which approach is better? Why are they different?
+1. Should we prefer `kubectl create ...` or `kubectl apply ...`? The examples
+   in the  [prior art](#prior-art) section use both.
+1. From what I understand, Vector requires the Kubernetes `watch` verb in order
+   to receive updates to k8s cluster changes. This is required for the
+   `kubernetes_pod_metadata` transform. Yet, Fluentbit [requires the `get`,
+   `list`, and `watch` verbs][fluentbit_role]. Why don't we require the same?
+1. What is `updateStrategy` ... `RollingUpdate`? This is not included in
+   [our daemonset][vector_daemonset] or in [any of Fluentbit's config
+   files][fluentbit_installation]. But it is included in both [Fluentd's
+   daemonset][fluentd_daemonset] and [LogDNA's daemonset][logdna_daemonset].
+1. I've also noticed `resources` declarations in some of these config files.
```

Comment:
@Alexx-G you're correct here. We have some understanding of what limitations make sense for vector, based on how we engineer it, for instance up until lately it didn't scale very well beyond 4 CPUs. We'll do our best to come up with sane resource limitations.

That said, it would make a lot of sense for a mindful operator to tweak the defaults that we'll provide for the best results on their cluster, and we'll have to provide guidance for that.

It would be great to publish a regularly updated bulletin on vector's runtime properties (i.e. performance considerations like how much memory/CPU we can utilize and under what conditions). That would be a real killer feature for everyone that wants to deploy vector under load, not just k8s.
Though it's a lot of additional hard work to determine these properties, people with large deployments tend to do this anyway to gain confidence in their setup. We could exchange this data with our partners and derive an even more realistic profile for Vector's runtime properties, having multiple data sets. This is worth a separate RFC though.

Reply:
Added a section on this to the RFC